// langchain/dist/chains/conversation.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConversationChain = void 0;
const llm_chain_js_1 = require("./llm_chain.cjs");
const index_js_1 = require("../prompts/index.cjs");
const index_js_2 = require("../memory/index.cjs");
const defaultTemplate = `The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
{history}
Human: {input}
AI:`;
class ConversationChain extends llm_chain_js_1.LLMChain {
    constructor(fields) {
        super({
            prompt: fields.prompt ??
                new index_js_1.PromptTemplate({
                    template: defaultTemplate,
                    inputVariables: ["history", "input"],
                }),
            llm: fields.llm,
            outputKey: fields.outputKey ?? "response",
        });
        this.memory = fields.memory ?? new index_js_2.BufferMemory();
    }
}
exports.ConversationChain = ConversationChain;
